src/Spring/Spring.Aop/Aspects/Cache/CacheResultAdvice.cs

```diff
@@ -78,7 +78,14 @@ private CacheResultInfo GetCacheResultInfo(MethodInfo method)
                 CacheResultItemsAttribute[] itemInfoArray = (CacheResultItemsAttribute[])GetCustomAttributes(method, typeof(CacheResultItemsAttribute));
 
                 cacheResultInfo = new CacheResultInfo(resultInfo, itemInfoArray);
-                _cacheResultAttributeCache[method] = cacheResultInfo;
+                // need lock! Hashtable support only one writer!
+                lock (_cacheResultAttributeCache)
+                {
+                    if (!_cacheResultAttributeCache.Contains(method))
+                    {
+                        _cacheResultAttributeCache[method] = cacheResultInfo;

```

Comment:
If you want this to be efficient it should probably use double-checked locking and thus move also

```csharp
CacheResultAttribute resultInfo = (CacheResultAttribute)GetCustomAttribute(method, typeof(CacheResultAttribute));
CacheResultItemsAttribute[] itemInfoArray = (CacheResultItemsAttribute[])GetCustomAttributes(method, typeof(CacheResultItemsAttribute));

cacheResultInfo = new CacheResultInfo(resultInfo, itemInfoArray);
```
inside the !Contains block. If you want to walk the extra mile, there's also the possibility to switch to Dictionary<TKey, TValue> and use TryGetValue instead of `if (cacheResultInfo == null)`

Thoughts?

Reply:
Or directly `ConcurrentDictionary<TKey,TValue>` available since .NET 4.0